### DISCUSSION TITLE
Before the plugin system

### DISCUSSION THREAD

**Discussion by n6646312** [helix-editor/helix#10291] [Q&A]
I understand that the devs have decided to adopt an integrated plugin ecosystem based on some dialect of Lisp (#8675), but it seems like it still takes some time to be ready. Until then, how much more extensibility can we expect?

I've noticed that Helix provides a variety of commands to interact with the shell (`:pipe`, `:insert-output`. `shell_keep_pipe`, etc.) and that there are PRs aimed at improving this (#6979, #8636). Will this trend continue? Or are there any attempts to introduce a temporary system that provides at least some degree of extensibility before the plugin system is ready?

**Comment by agieocean** [comment]
I'm working on an intermediate solution that uses generic python without any requirements (which means I can easily bundle it to a binary or even rewrite it in another language) and tmux send keys to automate helix for session management but there's 0 reason it couldn't be extended for other uses if there's interest. I know having to use tmux might be an issue for some people but this might be a good middle ground while the devs of helix work on figuring out what's best for the official plugin system.

At the moment the code supports loading, modifying and exporting (if you use the pr for command expansion) workspaces (just lists of open buffers atm). More pertinent though I just finished code for loading and switching layouts via splits including loading specific files to each window which requires using both commands and keyboard shortcuts and it works well. 

In my mind that means we can automate helix mostly and do a lot of what's required for plugins in whatever language people prefer while we wait but I'm interested in what other people think.

**Comment by L-Trump** [comment]
Maybe the thing we need before an entire plugin system is a "remote control" implementation for helix. Just allow other programs to read and edit file content/current configs/history/etc and invoke helix commands or functions through socket/fifo.

**Reply by agieocean** [reply]
I think that's a good idea and I think this may be a good way to add additional accessibility to helix, if I can find some time I'll look at adding this to the codebase

**Reply by MarSik** [reply]
What about starting with something like :pipe-cmd (synchronous)? The stdout of the invocation could be interpreted as a list of helix commands.

**Comment by Ben-PH** [comment]
How about a `wasm` or `wasi` approach? The idea being that you are not restricted to a single language, as SDKs for any `wasi/wasm` supported language could work...
